In [2]:
import pandas as pd
import numpy as np
import recordlinkage as rl
import re

# POKEMON

In [2]:
dex_scrap = pd.read_json('Scraping_datasets/pokemondb/pokedex.json')
dex_scrap.tail()

#         Name   Type1     Type2  Total   HP  Attack  Defense  \
1070  902  Basculegion   Water     Ghost    530  120      92       65   
1071  903     Sneasler  Poison  Fighting    510   80     130       60   
1072  904     Overqwil    Dark    Poison    510   85     115       95   
1073  905     Enamorus   Fairy    Flying    580   74     115       70   
1074  905     Enamorus   Fairy    Flying    580   74     115      110   

      Sp. Atk  Sp. Def  Speed                       sprite_url  \
1070      100       75     78  https://img.pokemondb.net/s.png   
1071       40       80    120  https://img.pokemondb.net/s.png   
1072       65       65     85  https://img.pokemondb.net/s.png   
1073      135       80    106  https://img.pokemondb.net/s.png   
1074      135      100     46  https://img.pokemondb.net/s.png   

              Variant  
1070           Female  
1071              NaN  
1072              NaN  
1073  Incarnate Forme  
1074    Therian Forme

In [3]:
species = pd.read_json('API_per_integrazione/species.json.gz', compression='gzip', orient='records', lines=True)
species.tail()

ID       Name                                          Varieties  \
893  894  Regieleki  [{'is_default': True, 'pokemon': {'name': 'reg...   
894  895  Regidrago  [{'is_default': True, 'pokemon': {'name': 'reg...   
895  896  Glastrier  [{'is_default': True, 'pokemon': {'name': 'gla...   
896  897  Spectrier  [{'is_default': True, 'pokemon': {'name': 'spe...   
897  898    Calyrex  [{'is_default': True, 'pokemon': {'name': 'cal...   

     Generation Evolves_from  Has_gender_diff     Rarity  
893           8         None            False  Legendary  
894           8         None            False  Legendary  
895           8         None            False  Legendary  
896           8         None            False  Legendary  
897           8         None            False  Legendary

In [4]:
#Inizializzo l'indexer per l'integrazione
indexer = rl.Index()
indexer.full()

<Index>

In [5]:
#Valuto il numero di coppie candidate al linking
candidates = indexer.index(species, dex_scrap)
print(len(candidates))

965350


In [6]:
compare = rl.Compare()
compare.string('Name', 'Name', threshold=0.85, label='Name')
compare.exact('ID', '#', label='ID')

features = compare.compute(candidates, species, dex_scrap)

In [7]:
features.sum(axis=1).value_counts().sort_index(ascending=False)

2.0      1054
1.0        23
0.0    964273
dtype: int64

Abbiamo 1054 match completi (nome+ID) e 23 match singoli (o nome o ID)

In [8]:
potential_matches = features[features.sum(axis=1) > 0].reset_index()
potential_matches['Score'] = potential_matches.loc[:, 'Name':'ID'].sum(axis=1)
potential_matches

level_0  level_1  Name  ID  Score
0           0        0   1.0   1    2.0
1           1        1   1.0   1    2.0
2           2        2   1.0   1    2.0
3           2        3   1.0   1    2.0
4           3        4   1.0   1    2.0
...       ...      ...   ...  ..    ...
1072      895     1061   1.0   1    2.0
1073      896     1062   1.0   1    2.0
1074      897     1063   1.0   1    2.0
1075      897     1064   1.0   1    2.0
1076      897     1065   1.0   1    2.0

[1077 rows x 5 columns]

In [9]:
potential_matches[potential_matches['Score'] == 1.0]

level_0  level_1  Name  ID  Score
40         28       40   0.0   1    1.0
42         29       44   1.0   0    1.0
44         31       43   0.0   1    1.0
45         32       41   1.0   0    1.0
161       118      355   1.0   0    1.0
164       121      161   0.0   1    1.0
165       121      162   0.0   1    1.0
184       136      292   1.0   0    1.0
285       221     1022   1.0   0    1.0
297       232      180   1.0   0    1.0
361       288      158   1.0   0    1.0
541       438      534   0.0   1    1.0
700       571      693   1.0   0    1.0
701       572      692   1.0   0    1.0
808       668      799   0.0   1    1.0
934       771      925   0.0   1    1.0
948       784      939   0.0   1    1.0
949       785      940   0.0   1    1.0
950       786      941   0.0   1    1.0
951       787      942   0.0   1    1.0
1031      863      279   1.0   0    1.0
1032      863      280   1.0   0    1.0
1035      865     1024   0.0   1    1.0

Analizzo manualmente i match con score pari a 1 (probabilmente basterebbe alzare la soglia, ma controllo ugualmente):

In [10]:
#Match numero 1: Nidoran femmina ---> VA TENUTO
print(species.loc[28])
print(dex_scrap.loc[40])

ID                                                                29
Name                                                       Nidoran-f
Varieties          [{'is_default': True, 'pokemon': {'name': 'nid...
Generation                                                         1
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 28, dtype: object
#                                                            29
Name                                                   Nidoran♀
Type1                                                    Poison
Type2                                                       NaN
Total                                                       275
HP                                                           55
Attack                                                       47
Defense                                      

In [11]:
#Match numero 2: Nidorino con Nidorina ---> ELIMINATO
print(species.loc[29])
print(dex_scrap.loc[44])

ID                                                                30
Name                                                        Nidorina
Varieties          [{'is_default': True, 'pokemon': {'name': 'nid...
Generation                                                         1
Evolves_from                                               Nidoran-f
Has_gender_diff                                                False
Rarity                                                        Common
Name: 29, dtype: object
#                                                            33
Name                                                   Nidorino
Type1                                                    Poison
Type2                                                       NaN
Total                                                       365
HP                                                           61
Attack                                                       72
Defense                                      

In [12]:
#Match numero 3: Nidoran maschio ---> VA TENUTO
print(species.loc[31])
print(dex_scrap.loc[43])

ID                                                                32
Name                                                       Nidoran-m
Varieties          [{'is_default': True, 'pokemon': {'name': 'nid...
Generation                                                         1
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 31, dtype: object
#                                                            32
Name                                                   Nidoran♂
Type1                                                    Poison
Type2                                                       NaN
Total                                                       273
HP                                                           46
Attack                                                       57
Defense                                      

In [13]:
#Match numero 4: Nidorino con Nidorina ---> ELIMINATO
print(species.loc[32])
print(dex_scrap.loc[41])

ID                                                                33
Name                                                        Nidorino
Varieties          [{'is_default': True, 'pokemon': {'name': 'nid...
Generation                                                         1
Evolves_from                                               Nidoran-m
Has_gender_diff                                                False
Rarity                                                        Common
Name: 32, dtype: object
#                                                            30
Name                                                   Nidorina
Type1                                                    Poison
Type2                                                       NaN
Total                                                       365
HP                                                           70
Attack                                                       62
Defense                                      

In [14]:
#Match numero 5: Slaking e Seaking sono diversi ---> ELIMINATO
print(species.loc[118])
print(dex_scrap.loc[355])

ID                                                               119
Name                                                         Seaking
Varieties          [{'is_default': True, 'pokemon': {'name': 'sea...
Generation                                                         1
Evolves_from                                                 Goldeen
Has_gender_diff                                                 True
Rarity                                                        Common
Name: 118, dtype: object
#                                                           289
Name                                                    Slaking
Type1                                                    Normal
Type2                                                       NaN
Total                                                       670
HP                                                          150
Attack                                                      160
Defense                                     

In [15]:
#Match numero 6: Mr Mime ---> VA TENUTO
print(species.loc[121])
print(dex_scrap.loc[161])

ID                                                               122
Name                                                         Mr-mime
Varieties          [{'is_default': True, 'pokemon': {'name': 'mr-...
Generation                                                         1
Evolves_from                                                 Mime-jr
Has_gender_diff                                                False
Rarity                                                        Common
Name: 121, dtype: object
#                                                           122
Name                                                        Mr.
Type1                                                   Psychic
Type2                                                     Fairy
Total                                                       460
HP                                                           40
Attack                                                       45
Defense                                     

In [16]:
#Match numero 7: Mr Mime ---> VA TENUTO
print(species.loc[121])
print(dex_scrap.loc[162])

ID                                                               122
Name                                                         Mr-mime
Varieties          [{'is_default': True, 'pokemon': {'name': 'mr-...
Generation                                                         1
Evolves_from                                                 Mime-jr
Has_gender_diff                                                False
Rarity                                                        Common
Name: 121, dtype: object
#                                                           122
Name                                                        Mr.
Type1                                                       Ice
Type2                                                   Psychic
Total                                                       460
HP                                                           50
Attack                                                       65
Defense                                     

In [17]:
#Match numero 8: Porygon e Porygon2 sono diversi ---> ELIMINATO
print(species.loc[136])
print(dex_scrap.loc[292])

ID                                                               137
Name                                                         Porygon
Varieties          [{'is_default': True, 'pokemon': {'name': 'por...
Generation                                                         1
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 136, dtype: object
#                                                           233
Name                                                   Porygon2
Type1                                                    Normal
Type2                                                       NaN
Total                                                       515
HP                                                           85
Attack                                                       80
Defense                                     

In [18]:
#Match numero 9: Corsola e Cursola sono diversi ---> ELIMINATO
print(species.loc[221])
print(dex_scrap.loc[1022])

ID                                                               222
Name                                                         Corsola
Varieties          [{'is_default': True, 'pokemon': {'name': 'cor...
Generation                                                         2
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 221, dtype: object
#                                                           864
Name                                                    Cursola
Type1                                                     Ghost
Type2                                                       NaN
Total                                                       510
HP                                                           60
Attack                                                       95
Defense                                     

In [19]:
#Match numero 10: Porygon e Porygon2 sono diversi ---> ELIMINATO
print(species.loc[232])
print(dex_scrap.loc[180])

ID                                                               233
Name                                                        Porygon2
Varieties          [{'is_default': True, 'pokemon': {'name': 'por...
Generation                                                         2
Evolves_from                                                 Porygon
Has_gender_diff                                                False
Rarity                                                        Common
Name: 232, dtype: object
#                                                           137
Name                                                    Porygon
Type1                                                    Normal
Type2                                                       NaN
Total                                                       395
HP                                                           65
Attack                                                       60
Defense                                     

In [20]:
#Match numero 11: Slaking e Seaking sono diversi ---> ELIMINATO
print(species.loc[288])
print(dex_scrap.loc[158])

ID                                                               289
Name                                                         Slaking
Varieties          [{'is_default': True, 'pokemon': {'name': 'sla...
Generation                                                         3
Evolves_from                                                Vigoroth
Has_gender_diff                                                False
Rarity                                                        Common
Name: 288, dtype: object
#                                                           119
Name                                                    Seaking
Type1                                                     Water
Type2                                                       NaN
Total                                                       450
HP                                                           80
Attack                                                       92
Defense                                     

In [21]:
#Match numero 12: Mime Jr è uguale ---> VA TENUTO
print(species.loc[438])
print(dex_scrap.loc[534])

ID                                                               439
Name                                                         Mime-jr
Varieties          [{'is_default': True, 'pokemon': {'name': 'mim...
Generation                                                         4
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                          Baby
Name: 438, dtype: object
#                                                           439
Name                                                       Mime
Type1                                                   Psychic
Type2                                                     Fairy
Total                                                       310
HP                                                           20
Attack                                                       25
Defense                                     

In [22]:
#Match numero 13 e 14: Minccino e Cinccino sono diversi ---> ELIMINATO
print(species.loc[571])
print(dex_scrap.loc[693])

print(species.loc[572])
print(dex_scrap.loc[692])

ID                                                               572
Name                                                        Minccino
Varieties          [{'is_default': True, 'pokemon': {'name': 'min...
Generation                                                         5
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 571, dtype: object
#                                                           573
Name                                                   Cinccino
Type1                                                    Normal
Type2                                                       NaN
Total                                                       470
HP                                                           75
Attack                                                       95
Defense                                     

In [23]:
#Match numero 15: C'è solo l'accento sbagliato ---> VA TENUTO
print(species.loc[668])
print(dex_scrap.loc[799])

ID                                                               669
Name                                                         Flabebe
Varieties          [{'is_default': True, 'pokemon': {'name': 'fla...
Generation                                                         6
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 668, dtype: object
#                                                           669
Name                                                    Flabébé
Type1                                                     Fairy
Type2                                                       NaN
Total                                                       303
HP                                                           44
Attack                                                       38
Defense                                     

In [24]:
#Match numero 16: Type NULL e Type: sono uguali ---> VA TENUTO
print(species.loc[771])
print(dex_scrap.loc[925])

ID                                                               772
Name                                                       Type-null
Varieties          [{'is_default': True, 'pokemon': {'name': 'typ...
Generation                                                         7
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                        Common
Name: 771, dtype: object
#                                                           772
Name                                                      Type:
Type1                                                    Normal
Type2                                                       NaN
Total                                                       534
HP                                                           95
Attack                                                       95
Defense                                     

In [25]:
#Match numero 17, 18, 19 e 20: I Tapu sono uguali ---> TENERE
print(species.loc[784])
print(dex_scrap.loc[939])

print(species.loc[785])
print(dex_scrap.loc[940])

print(species.loc[786])
print(dex_scrap.loc[941])

print(species.loc[787])
print(dex_scrap.loc[942])

ID                                                               785
Name                                                       Tapu-koko
Varieties          [{'is_default': True, 'pokemon': {'name': 'tap...
Generation                                                         7
Evolves_from                                                    None
Has_gender_diff                                                False
Rarity                                                     Legendary
Name: 784, dtype: object
#                                                           785
Name                                                       Tapu
Type1                                                  Electric
Type2                                                     Fairy
Total                                                       570
HP                                                           70
Attack                                                      115
Defense                                     

In [26]:
#Match numero 21 e 22: Cursola e Corsola sono diversi ---> ELIMINARE
print(species.loc[863])
print(dex_scrap.loc[279])

print(species.loc[863])
print(dex_scrap.loc[280])

ID                                                               864
Name                                                         Cursola
Varieties          [{'is_default': True, 'pokemon': {'name': 'cur...
Generation                                                         8
Evolves_from                                                 Corsola
Has_gender_diff                                                False
Rarity                                                        Common
Name: 863, dtype: object
#                                                           222
Name                                                    Corsola
Type1                                                     Water
Type2                                                      Rock
Total                                                       410
HP                                                           65
Attack                                                       55
Defense                                     

In [27]:
#Match numero 21 e 22: Sono lo stesso ---> TENERE
print(species.loc[865])
print(dex_scrap.loc[1024])

ID                                                               866
Name                                                         Mr-rime
Varieties          [{'is_default': True, 'pokemon': {'name': 'mr-...
Generation                                                         8
Evolves_from                                                 Mr-mime
Has_gender_diff                                                False
Rarity                                                        Common
Name: 865, dtype: object
#                                                           866
Name                                                        Mr.
Type1                                                       Ice
Type2                                                   Psychic
Total                                                       520
HP                                                           80
Attack                                                       85
Defense                                     

Si nota come nei match dove lo score è 1 (quindi non coincide il nome o non coincide l'ID) sono da scartare i match dove coincide il nome ma non l'ID (la coincidenza del nome viene dal fatto che sono nomi simili, dove magari cambia solo una lettera come nel caso "SLAKING-SEAKING") mentre sono da tenere i match dove non c'è coincidenza nel nome ma c'è coincidenza nell'ID. Quindi basta droppare queste righe:

In [28]:
potential_matches.drop(potential_matches[potential_matches.ID == 0].index, inplace=True)

In [29]:
potential_matches.head()

level_0  level_1  Name  ID  Score
0        0        0   1.0   1    2.0
1        1        1   1.0   1    2.0
2        2        2   1.0   1    2.0
3        2        3   1.0   1    2.0
4        3        4   1.0   1    2.0

In [30]:
species.head()

ID        Name                                          Varieties  \
0   1   Bulbasaur  [{'is_default': True, 'pokemon': {'name': 'bul...   
1   2     Ivysaur  [{'is_default': True, 'pokemon': {'name': 'ivy...   
2   3    Venusaur  [{'is_default': True, 'pokemon': {'name': 'ven...   
3   4  Charmander  [{'is_default': True, 'pokemon': {'name': 'cha...   
4   5  Charmeleon  [{'is_default': True, 'pokemon': {'name': 'cha...   

   Generation Evolves_from  Has_gender_diff  Rarity  
0           1         None            False  Common  
1           1    Bulbasaur            False  Common  
2           1      Ivysaur             True  Common  
3           1         None            False  Common  
4           1   Charmander            False  Common

Procedo con il merging:

In [31]:
#Aggiungo le informazioni da species (la colonna riferita a species è level_0)
pokemon_complete = pd.merge(potential_matches, species, left_on="level_0", right_index=True)

In [32]:
#Aggiungo le informazioni da scraping (la colonna riferita a dex_scrap è level_1)
pokemon_complete = pd.merge(pokemon_complete, dex_scrap, left_on="level_1", right_index=True)

In [33]:
pokemon_complete.drop(["level_0", "level_1", "Name_x", "ID_x", "Score", "ID_y", "Name_y"], axis = 1, inplace=True)

In [34]:
pokemon_complete = pokemon_complete.reindex(columns=['#', 'Name', 'Generation', 'Rarity', 'Evolves_from', 'Has_gender_diff',
                                                    'Type1', 'Type2', 'Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def',
                                                    'Speed', 'Variant', 'Varieties', 'sprite_url'])
pokemon_complete.reset_index(inplace=True)

In [35]:
pokemon_complete

index    #        Name  Generation     Rarity Evolves_from  \
0         0    1   Bulbasaur           1     Common         None   
1         1    2     Ivysaur           1     Common    Bulbasaur   
2         2    3    Venusaur           1     Common      Ivysaur   
3         3    3    Venusaur           1     Common      Ivysaur   
4         4    4  Charmander           1     Common         None   
...     ...  ...         ...         ...        ...          ...   
1061   1072  896   Glastrier           8  Legendary         None   
1062   1073  897   Spectrier           8  Legendary         None   
1063   1074  898     Calyrex           8  Legendary         None   
1064   1075  898     Calyrex           8  Legendary         None   
1065   1076  898     Calyrex           8  Legendary         None   

      Has_gender_diff    Type1   Type2  Total   HP  Attack  Defense  Sp. Atk  \
0               False    Grass  Poison    318   45      49       49       65   
1               False    Grass  Poison    405   60      62       63       80   
2                True    Grass  Poison    525   80      82       83      100   
3                True    Grass  Poison    625   80     100      123      122   
4               False     Fire     NaN    309   39      52       43       60   
...               ...      ...     ...    ...  ...     ...      ...      ...   
1061            False      Ice     NaN    580  100     145      130       65   
1062            False    Ghost     NaN    580  100      65       60      145   
1063            False  Psychic   Grass    500  100      80       80       80   
1064            False  Psychic     Ice    680  100     165      150       85   
1065            False  Psychic   Ghost    680  100      85       80      165   

      Sp. Def  Speed        Variant  \
0          65     45            NaN   
1          80     60            NaN   
2         100     80            NaN   
3         120     80  Mega Venusaur   
4          50     65            NaN   
...       ...    ...            ...   
1061      110     30            NaN   
1062       80    130            NaN   
1063       80     80            NaN   
1064      130     50      Ice Rider   
1065      100    150   Shadow Rider   

                                              Varieties  \
0     [{'is_default': True, 'pokemon': {'name': 'bul...   
1     [{'is_default': True, 'pokemon': {'name': 'ivy...   
2     [{'is_default': True, 'pokemon': {'name': 'ven...   
3     [{'is_default': True, 'pokemon': {'name': 'ven...   
4     [{'is_default': True, 'pokemon': {'name': 'cha...   
...                                                 ...   
1061  [{'is_default': True, 'pokemon': {'name': 'gla...   
1062  [{'is_default': True, 'pokemon': {'name': 'spe...   
1063  [{'is_default': True, 'pokemon': {'name': 'cal...   
1064  [{'is_default': True, 'pokemon': {'name': 'cal...   
1065  [{'is_default': True, 'pokemon': {'name': 'cal...   

                                             sprite_url  
0     https://img.pokemondb.net/sprites/sword-shield...  
1     https://img.pokemondb.net/sprites/sword-shield...  
2     https://img.pokemondb.net/sprites/sword-shield...  
3     https://img.pokemondb.net/sprites/sword-shield...  
4     https://img.pokemondb.net/sprites/sword-shield...  
...                                                 ...  
1061  https://img.pokemondb.net/sprites/sword-shield...  
1062  https://img.pokemondb.net/sprites/sword-shield...  
1063  https://img.pokemondb.net/sprites/sword-shield...  
1064  https://img.pokemondb.net/sprites/sword-shield...  
1065  https://img.pokemondb.net/sprites/sword-shield...  

[1066 rows x 19 columns]

# ITEMS

In [3]:
item_scrap = pd.read_json('Scraping_datasets/pokemondb/items.json')
item_scrap.head()

Name       Category  \
0  Ability Capsule     Hold items   
1    Ability Patch  General items   
2     Ability Urge   Battle items   
3        Abomasite     Hold items   
4         Absolite     Hold items   

                                              Effect  \
0  A capsule that allows a Pokémon with two Abili...   
1  Changes a Pokémon's ability to its Hidden Abil...   
2  When used, it activates the Ability of an ally...   
3    Enables Abomasnow to Mega Evolve during battle.   
4        Enables Absol to Mega Evolve during battle.   

                                          sprite_url  
0  https://img.pokemondb.net/sprites/items/abilit...  
1                    https://img.pokemondb.net/s.png  
2  https://img.pokemondb.net/sprites/items/abilit...  
3  https://img.pokemondb.net/sprites/items/abomas...  
4  https://img.pokemondb.net/sprites/items/absoli...

In [7]:
item_api = pd.read_json('API_per_integrazione/items.json.gz', compression='gzip', orient='records', lines=True)
item_api.head()

ID         Name                                         Attributes  \
0   1  Master-ball  [countable, consumable, usable-in-battle, hold...   
1   2   Ultra-ball  [countable, consumable, usable-in-battle, hold...   
2   3   Great-ball  [countable, consumable, usable-in-battle, hold...   
3   4    Poke-ball  [countable, consumable, usable-in-battle, hold...   
4   5  Safari-ball  [countable, consumable, usable-in-battle, hold...   

         Category                                             Effect  
0  Standard-balls                 Catches a wild Pokémon every time.  
1  Standard-balls  Tries to catch a wild Pokémon.  Success rate i...  
2  Standard-balls  Tries to catch a wild Pokémon.  Success rate i...  
3  Standard-balls                     Tries to catch a wild Pokémon.  
4  Standard-balls  Tries to catch a wild Pokémon in the Great Mar...

In [8]:
#Tolgo tutti i - e impongo che tutte le parole abbiano la prima lettera maisucola, in modo tale da facilitare il matching

#Dataset API
item_api['Name'] = item_api['Name'].str.replace('-',' ')
item_api['Name'] = item_api['Name'].str.title()

item_api['Category'] = item_api['Category'].str.replace('-',' ')
item_api['Category'] = item_api['Category'].str.title()

item_api.sort_values('Name', inplace=True)
item_api.reset_index(inplace=True)

item_api.drop(['index', 'ID'], axis=1, inplace=True)
item_api.head()

Name Attributes         Category  \
0  Ability Capsule         []         Vitamins   
1    Ability Patch         []         Vitamins   
2     Ability Urge         []  Miracle Shooter   
3        Abomasite         []      Mega Stones   
4       Abra Candy         []  Species Candies   

                                              Effect  
0  Switches a Pokémon between its two possible (n...  
1                                               None  
2   Forcibly activates a friendly Pokémon's ability.  
3  Held: Allows Abomasnow to Mega Evolve into Meg...  
4                                               None

In [9]:
#Dataset scraping
item_scrap['Name'] = item_scrap['Name'].str.replace('-',' ')
item_scrap['Name'] = item_scrap['Name'].str.title()
item_scrap.head()

Name       Category  \
0  Ability Capsule     Hold items   
1    Ability Patch  General items   
2     Ability Urge   Battle items   
3        Abomasite     Hold items   
4         Absolite     Hold items   

                                              Effect  \
0  A capsule that allows a Pokémon with two Abili...   
1  Changes a Pokémon's ability to its Hidden Abil...   
2  When used, it activates the Ability of an ally...   
3    Enables Abomasnow to Mega Evolve during battle.   
4        Enables Absol to Mega Evolve during battle.   

                                          sprite_url  
0  https://img.pokemondb.net/sprites/items/abilit...  
1                    https://img.pokemondb.net/s.png  
2  https://img.pokemondb.net/sprites/items/abilit...  
3  https://img.pokemondb.net/sprites/items/abomas...  
4  https://img.pokemondb.net/sprites/items/absoli...

In [11]:
#Inizializzo l'indexer per l'integrazione e il matching
indexer = rl.Index()
indexer.full()

<Index>

In [12]:
candidates = indexer.index(item_api, item_scrap)
print(len(candidates))

1542720


In [13]:
#Voglio tenere tutti i match esatti fra gli strumenti, ma tengo anche quelli molto simili
compare = rl.Compare()
compare.string('Name', 'Name', threshold=0.9, label='Name')
compare.exact('Name', 'Name', label='Exact')

features = compare.compute(candidates, item_api, item_scrap)

In [14]:
features.sum(axis=1).value_counts().sort_index(ascending=False)

2.0        826
1.0         42
0.0    1541852
dtype: int64

Ci sono 826 match esatti, 42 match potenziali (stringhe di testo molto simili ma non esattamente identiche)

In [15]:
potential_matches = features[features.sum(axis=1) > 0].reset_index()
potential_matches['Score'] = potential_matches.loc[:, 'Name':'Exact'].sum(axis=1)
potential_matches.head()

level_0  level_1  Name  Exact  Score
0        0        0   1.0      1    2.0
1        1        1   1.0      1    2.0
2        2        2   1.0      1    2.0
3        3        3   1.0      1    2.0
4        5        4   1.0      1    2.0

In [16]:
#Suddivido i match esatti e quelli simili
exact_matches = potential_matches[potential_matches['Score'] == 2.0]
temp = potential_matches[potential_matches['Score'] == 1.0]

### Tratto i match potenziali

In [17]:
#La soglia l'ho posta molto alta per non avere un numero eccessivo di match potenziali
item_similar = pd.merge(temp, item_api, left_on="level_0", right_index=True)
item_similar = pd.merge(item_similar, item_scrap, left_on="level_1", right_index=True)
item_similar.reset_index(inplace=True)
item_similar.head(50)

index  level_0  level_1  Name_x  Exact  Score           Name_y  \
0      75      128       94     1.0      0    1.0   Charizardite X   
1      76      129       93     1.0      0    1.0   Charizardite Y   
2     115      218      138     1.0      0    1.0       Dire Hit 2   
3     116      219      137     1.0      0    1.0       Dire Hit 3   
4     153      582      179     1.0      0    1.0      Exp Candy L   
5     154      583      180     1.0      0    1.0      Exp Candy M   
6     155      584      181     1.0      0    1.0      Exp Candy S   
7     156      585      182     1.0      0    1.0     Exp Candy Xl   
8     157      586      183     1.0      0    1.0     Exp Candy Xs   
9     158      588      184     1.0      0    1.0        Exp Share   
10    227      700      269     1.0      0    1.0       Guard Spec   
11    343      909      411     1.0      0    1.0     Mewtwonite X   
12    344      910      410     1.0      0    1.0     Mewtwonite Y   
13    489     1145      578     1.0      0    1.0       Roto Catch   
14    491     1147      576     1.0      0    1.0  Roto Exp Points   
15    493     1149      574     1.0      0    1.0       Roto Hatch   
16    496     1150      580     1.0      0    1.0  Roto Hp Restore   
17    497     1151      579     1.0      0    1.0  Roto Pp Restore   
18    821     1566      929     1.0      0    1.0     X Accuracy 2   
19    827     1568      929     1.0      0    1.0     X Accuracy 6   
20    822     1566      930     1.0      0    1.0     X Accuracy 2   
21    825     1567      930     1.0      0    1.0     X Accuracy 3   
22    823     1567      928     1.0      0    1.0     X Accuracy 3   
23    826     1568      928     1.0      0    1.0     X Accuracy 6   
24    831     1570      933     1.0      0    1.0       X Attack 2   
25    837     1572      933     1.0      0    1.0       X Attack 6   
26    832     1570      934     1.0      0    1.0       X Attack 2   
27    835     1571      934     1.0      0    1.0       X Attack 3   
28    833     1571      932     1.0      0    1.0       X Attack 3   
29    836     1572      932     1.0      0    1.0       X Attack 6   
30    841     1574      937     1.0      0    1.0      X Defense 2   
31    847     1576      937     1.0      0    1.0      X Defense 6   
32    842     1574      938     1.0      0    1.0      X Defense 2   
33    845     1575      938     1.0      0    1.0      X Defense 3   
34    843     1575      936     1.0      0    1.0      X Defense 3   
35    846     1576      936     1.0      0    1.0      X Defense 6   
36    849     1578      940     1.0      0    1.0       X Sp Atk 2   
37    850     1579      941     1.0      0    1.0       X Sp Atk 3   
38    851     1580      942     1.0      0    1.0       X Sp Atk 6   
39    852     1582      944     1.0      0    1.0       X Sp Def 2   
40    853     1583      945     1.0      0    1.0       X Sp Def 3   
41    854     1584      946     1.0      0    1.0       X Sp Def 6   

                                           Attributes       Category_x  \
0                                                  []      Mega Stones   
1                                                  []      Mega Stones   
2                                                  []  Miracle Shooter   
3                                                  []  Miracle Shooter   
4                                                  []         Vitamins   
5                                                  []         Vitamins   
6                                                  []         Vitamins   
7                                                  []         Vitamins   
8                                                  []         Vitamins   
9                         [holdable, holdable-active]         Training   
10  [countable, consumable, usable-in-battle, hold...      Stat Boosts   
11                                                 []      Mega Stones   
12                             

Guardando i nomi di questi match potenziali si vede che quelli che effettivamente coincidono sono quelli in cui da una parte abbiamo Exp, Spec, Sp, ... senza il . finale, mentre dall'altra abbiamo Exp., Spec., Sp., ... con il . finale

Vado a tenere questi e ad eliminare gli altri:

In [20]:
item_similar['Name_y'] = item_similar['Name_y'].str.replace('Exp','Exp.')
item_similar['Name_y'] = item_similar['Name_y'].str.replace('X Sp','X Sp.')
item_similar['Name_y'] = item_similar['Name_y'].str.replace('Spec','Spec.')

In [22]:
match = item_similar[item_similar['Name_y'] == item_similar['Name']]
match.head()

index  level_0  level_1  Name_x  Exact  Score         Name_y Attributes  \
4    153      582      179     1.0      0    1.0   Exp. Candy L         []   
5    154      583      180     1.0      0    1.0   Exp. Candy M         []   
6    155      584      181     1.0      0    1.0   Exp. Candy S         []   
7    156      585      182     1.0      0    1.0  Exp. Candy Xl         []   
8    157      586      183     1.0      0    1.0  Exp. Candy Xs         []   

  Category_x Effect_x           Name Category_y  \
4   Vitamins     None   Exp. Candy L   Medicine   
5   Vitamins     None   Exp. Candy M   Medicine   
6   Vitamins     None   Exp. Candy S   Medicine   
7   Vitamins     None  Exp. Candy Xl   Medicine   
8   Vitamins     None  Exp. Candy Xs   Medicine   

                                        Effect_y  \
4                                            NaN   
5                                            NaN   
6                                            NaN   
7  Increases the Pokémon's Exp points by 30,000.   
8                                            NaN   

                        sprite_url  
4  https://img.pokemondb.net/s.png  
5  https://img.pokemondb.net/s.png  
6  https://img.pokemondb.net/s.png  
7  https://img.pokemondb.net/s.png  
8  https://img.pokemondb.net/s.png

### Tratto i match esatti

In [23]:
#Aggiungo le informazioni da API (la colonna riferita a API è level_0)
item_complete = pd.merge(exact_matches, item_api, left_on="level_0", right_index=True)

#Aggiungo le informazioni da scraping (la colonna riferita a scraping è level_1)
item_complete = pd.merge(item_complete, item_scrap, left_on="level_1", right_index=True)

item_complete

level_0  level_1  Name_x  Exact  Score           Name_y  \
0          0        0     1.0      1    2.0  Ability Capsule   
1          1        1     1.0      1    2.0    Ability Patch   
2          2        2     1.0      1    2.0     Ability Urge   
3          3        3     1.0      1    2.0        Abomasite   
4          5        4     1.0      1    2.0         Absolite   
..       ...      ...     ...    ...    ...              ...   
863     1597      955     1.0      1    2.0     Yellow Scarf   
864     1598      956     1.0      1    2.0     Yellow Shard   
865     1601      957     1.0      1    2.0        Zap Plate   
866     1603      958     1.0      1    2.0             Zinc   
867     1604      959     1.0      1    2.0        Zoom Lens   

                                            Attributes       Category_x  \
0                                                   []         Vitamins   
1                                                   []         Vitamins   
2                                                   []  Miracle Shooter   
3                                                   []      Mega Stones   
4                                                   []      Mega Stones   
..                                                 ...              ...   
863                                  [holdable-active]          Scarves   
864                                      [underground]     Collectibles   
865           [holdable, holdable-active, underground]           Plates   
866  [countable, consumable, usable-overworld, usab...         Vitamins   
867                        [holdable, holdable-active]       Held Items   

                                              Effect_x             Name  \
0    Switches a Pokémon between its two possible (n...  Ability Capsule   
1                                                 None    Ability Patch   
2     Forcibly activates a friendly Pokémon's ability.     Ability Urge   
3    Held: Allows Abomasnow to Mega Evolve into Meg...        Abomasite   
4    Held: Allows Absol to Mega Evolve into Mega Ab...         Absolite   
..                                                 ...              ...   
863  Raises the holder's Toughness while in a contest.     Yellow Scarf   
864       No effect. Can be traded for items or moves.     Yellow Shard   
865  Held: Electric-Type moves from holder do 20% m...        Zap Plate   
866              Raises Special Defense and happiness.             Zinc   
867  Held: Provides a 1/5 (20%) boost in accuracy i...        Zoom Lens   

        Category_y                                           Effect_y  \
0       Hold items  A capsule that allows a Pokémon with two Abili...   
1    General items  Changes a Pokémon's ability to its Hidden Abil...   
2     Battle items  When used, it activates the Ability of an ally...   
3       Hold items    Enables Abomasnow to Mega Evolve during battle.   
4       Hold items        Enables Absol to Mega Evolve during battle.   
..             ...                                                ...   
863  General items         Raises holder's Tough aspect in a Contest.   
864  General items  A small yellow shard. It appears to be from so...   
865     Hold items  Increases power of Electric-type moves. Change...   
866       Medicine               Increases Special Defense EVs by 10.   
867     Hold items  Raises a move's accuracy if the holder moves a...   

                                            sprite_url  
0    https://img.pokemondb.net/sprites/items/abilit...  
1                      https://img.pokemondb.net/s.png  
2    https://img.pokemondb.net/sprites/items/abilit...  
3    https://img.pokemondb.net/sprites/items/abomas...  
4    https://img.pokemondb.net/sprites/items/absoli...  
..                                                 ...  
863  https://img.pokemondb.net/sprites/items/yellow...  
864  https://img.pokemondb.net/sprites/items/yellow...  
865  https://img.pokemondb.net/sprites/items

### Concatenazione match esatti e simili

In [24]:
item_total = pd.concat([item_complete, match])
item_total.sort_values('Name', inplace=True)
item_total.drop(["level_0", "level_1", "Name_x", "Exact", "Score", "Name_y", "index"], axis = 1, inplace=True)
item_total.tail()

Attributes    Category_x  \
863                                  [holdable-active]       Scarves   
864                                      [underground]  Collectibles   
865           [holdable, holdable-active, underground]        Plates   
866  [countable, consumable, usable-overworld, usab...      Vitamins   
867                        [holdable, holdable-active]    Held Items   

                                              Effect_x          Name  \
863  Raises the holder's Toughness while in a contest.  Yellow Scarf   
864       No effect. Can be traded for items or moves.  Yellow Shard   
865  Held: Electric-Type moves from holder do 20% m...     Zap Plate   
866              Raises Special Defense and happiness.          Zinc   
867  Held: Provides a 1/5 (20%) boost in accuracy i...     Zoom Lens   

        Category_y                                           Effect_y  \
863  General items         Raises holder's Tough aspect in a Contest.   
864  General items  A small yellow shard. It appears to be from so...   
865     Hold items  Increases power of Electric-type moves. Change...   
866       Medicine               Increases Special Defense EVs by 10.   
867     Hold items  Raises a move's accuracy if the holder moves a...   

                                            sprite_url  
863  https://img.pokemondb.net/sprites/items/yellow...  
864  https://img.pokemondb.net/sprites/items/yellow...  
865  https://img.pokemondb.net/sprites/items/zap-pl...  
866   https://img.pokemondb.net/sprites/items/zinc.png  
867  https://img.pokemondb.net/sprites/items/zoom-l...

In [25]:
item_total.reset_index(inplace=True)
item_total.drop(["index"], axis = 1, inplace=True)
item_total.tail()

Attributes    Category_x  \
835                                  [holdable-active]       Scarves   
836                                      [underground]  Collectibles   
837           [holdable, holdable-active, underground]        Plates   
838  [countable, consumable, usable-overworld, usab...      Vitamins   
839                        [holdable, holdable-active]    Held Items   

                                              Effect_x          Name  \
835  Raises the holder's Toughness while in a contest.  Yellow Scarf   
836       No effect. Can be traded for items or moves.  Yellow Shard   
837  Held: Electric-Type moves from holder do 20% m...     Zap Plate   
838              Raises Special Defense and happiness.          Zinc   
839  Held: Provides a 1/5 (20%) boost in accuracy i...     Zoom Lens   

        Category_y                                           Effect_y  \
835  General items         Raises holder's Tough aspect in a Contest.   
836  General items  A small yellow shard. It appears to be from so...   
837     Hold items  Increases power of Electric-type moves. Change...   
838       Medicine               Increases Special Defense EVs by 10.   
839     Hold items  Raises a move's accuracy if the holder moves a...   

                                            sprite_url  
835  https://img.pokemondb.net/sprites/items/yellow...  
836  https://img.pokemondb.net/sprites/items/yellow...  
837  https://img.pokemondb.net/sprites/items/zap-pl...  
838   https://img.pokemondb.net/sprites/items/zinc.png  
839  https://img.pokemondb.net/sprites/items/zoom-l...

In [26]:
item_total = item_total.reindex(columns=['Name', 'Effect_x', 'Effect_y', 'Attributes', 'Category_x', 'Category_y', 
                                         'sprite_url'])

item_total.rename(columns={"Name": "Name", "Effect_x": "Effect_API", "Effect_y": "Effect_SCRAPING", "Attributes": "Attributes",
                          "Category_x": "Category_API", "Category_y": "Category_SCRAPING", "sprite_url": "sprite_url"})

Name                                         Effect_API  \
0    Ability Capsule  Switches a Pokémon between its two possible (n...   
1      Ability Patch                                               None   
2       Ability Urge   Forcibly activates a friendly Pokémon's ability.   
3          Abomasite  Held: Allows Abomasnow to Mega Evolve into Meg...   
4           Absolite  Held: Allows Absol to Mega Evolve into Mega Ab...   
..               ...                                                ...   
835     Yellow Scarf  Raises the holder's Toughness while in a contest.   
836     Yellow Shard       No effect. Can be traded for items or moves.   
837        Zap Plate  Held: Electric-Type moves from holder do 20% m...   
838             Zinc              Raises Special Defense and happiness.   
839        Zoom Lens  Held: Provides a 1/5 (20%) boost in accuracy i...   

                                       Effect_SCRAPING  \
0    A capsule that allows a Pokémon with two Abili...   
1    Changes a Pokémon's ability to its Hidden Abil...   
2    When used, it activates the Ability of an ally...   
3      Enables Abomasnow to Mega Evolve during battle.   
4          Enables Absol to Mega Evolve during battle.   
..                                                 ...   
835         Raises holder's Tough aspect in a Contest.   
836  A small yellow shard. It appears to be from so...   
837  Increases power of Electric-type moves. Change...   
838               Increases Special Defense EVs by 10.   
839  Raises a move's accuracy if the holder moves a...   

                                            Attributes     Category_API  \
0                                                   []         Vitamins   
1                                                   []         Vitamins   
2                                                   []  Miracle Shooter   
3                                                   []      Mega Stones   
4                                                   []      Mega Stones   
..                                                 ...              ...   
835                                  [holdable-active]          Scarves   
836                                      [underground]     Collectibles   
837           [holdable, holdable-active, underground]           Plates   
838  [countable, consumable, usable-overworld, usab...         Vitamins   
839                        [holdable, holdable-active]       Held Items   

    Category_SCRAPING                                         sprite_url  
0          Hold items  https://img.pokemondb.net/sprites/items/abilit...  
1       General items                    https://img.pokemondb.net/s.png  
2        Battle items  https://img.pokemondb.net/sprites/items/abilit...  
3          Hold items  https://img.pokemondb.net/sprites/items/abomas...  
4          Hold items  https://img.pokemondb.net/sprites/items/absoli...  
..                ...                                                ...  
835     General items  https://img.pokemondb.net/sprites/items/yellow...  
836     General items  https://img.pokemondb.net/sprites/items/yellow...  
837        Hold items  https://img.pokemondb.net/sprites/items/zap-pl...  
838          Medicine   https://img.pokemondb.net/sprites/items/zinc.png  
839        Hold items  https://img.pokemondb.net/sprites/items/zoom-l...  

[840 rows x 7 columns]